README.md
Cross-Browser Rich Text Editor
==============================

The original Cross-Browser Rich Text Editor (RTE) is based on the designMode() functionality introduced in Internet Explorer 5, and implemented in Mozilla 1.3+ using the Mozilla Rich Text Editing API.

Project Home: http://www.kevinroth.com/rte


changelog.txt
Cross-Browser Rich Text Editor change log

3.14 - 1/10/2011 - fixed cleanWordContent function and updated to strip potentially malicious tags
3.14 - 1/10/2011 - renamed richtext_compressed.js to richtext.min.js
3.13 - 9/27/2010 - Updated html2xhtml.js file to keep correct case on special characters such as accented characters, etc
3.12a - 8/3/2010 - Fixed typo that corrects Safari support
3.12 - 7/30/2010 - Disabled rich text editor for iPads, as it is currently incompatible (thanks Charles)
3.11 - 7/27/2010 - Focus-setting for Gecko browsers was only necessary prior to v1.9 (Firefox 3) - added exception
3.11 - 7/27/2010 - Fixed issue with readOnly status and view source checkbox
3.10 - 7/22/2010 - Disabled rich text editor for iPhones, as it is currently incompatible
3.10 - 7/22/2010 - Fixed problem with get_webkit_version() function that may have caused iPhone's Safari to crash (thanks Dale)
3.10 - 7/22/2010 - Fixed readOnly option in Firefox/IE7 (thanks Alex)
3.09 - 7/6/2008 - Fixed Firefox tab replacement bug in code view
3.09 - 7/6/2008 - Fixed style code showing up in output in firefox 3
3.09 - 7/6/2008 - Fixed IE hang bug
3.08b - 9/6/2006 - Fixed local and external style references from being inserted in IE
3.08a - 9/4/2006 - Fixed bug where firefox was not recognizing external css file.
3.08 - 8/27/2006 - Fixed bug where view source checkbox stayed checked when refreshing browser
3.08 - 8/27/2006 - Added insert special character function
3.08 - 8/26/2006 - Improved palette size and uniformity across browsers
3.08 - 8/26/2006 - Added support for text color and background color in Safari
3.08 - 8/22/2006 - Added support for the next version of Safari, webkit version 420+ (experimental), in which all RTE commands should be supported.
3.08 - 8/21/2006 - Added support for cmdUnlink command (see usage documentation).
3.08 - 8/21/2006 - Fixed palette positioning.
3.08 - 8/18/2006 - Added parameter to initRTE() for encHTML, which is false by default.  User can now toggle html encoding by setting this parameter.  ASPX version requires this to be set to true.
3.07a - 8/5/2006 - Fixed htmlEncoding.  Content was actually being urlencoded, not htmlencoded.  htmlEncode() and htmlDecode() javascript functions are now included in the richtext.js/richtext_compressed files
3.07 - 8/5/2006 - Security Fix: Content posted from RTE is now htmlencoded.  When reading stored html data from a database, make sure to run htmldecode on your content string (RTE will automatically decode, however).
3.07 - 8/5/2006 - Fixed problem with cleanWordContent() function that was not allowing some commands to work in IE
3.07 - 8/5/2006 - Added ASP.NET (VB) demo
3.06 - 8/2/2006 - Added JSP demo to zip file
3.06 - 8/2/2006 - Introduced getHtmlSrc() and setHtmlSrc() functions for js developers to more easily access (and update ) html source of given rte
3.05a - 7/27/2006 - Fixed color palette
3.05a - 7/27/2006 - Fixed bug with relative links fix
3.05 - 7/26/2006 - Added cleanWordContent() function to clean nasty html pasted from word
3.05 - 7/26/2006 - Replaced Vyacheslav Smolin's HTML2XHTML code with Jacob Lee's.  The HTML2XHTML functionality is now free for noncommercial & commercial use (thanks Jacob Lee!)
3.05 - 7/26/2006 - Removed references to parent.command in favor of local variables (thanks Florian!)
3.05 - 7/26/2006 - Fixed RTE from adding http:// to relative links when inserting links to current domain
3.05 - 7/26/2006 - Enabled CSS support for firefox
3.05 - 7/26/2006 - Fixed insertURL function when selection text contains the "&" character
3.04 - 7/11/2006 - Added partial inserthtml functionality for Opera 9.  Inserting tables crashes browser, so disabled
3.04 - 7/11/2006 - Added Increase/Decrease font size support for Opera 9+
3.04 - 7/11/2006 - Fixed bug with Gecko browsers where could not insert lists if rte did not have focus
3.04 - 7/11/2006 - Fixed bug with IE where html could be inserted in main page content if rte did not have focus (thanks Derin Basden (aka Kirok)!)
3.04 - 7/11/2006 - Replaced trim() function to use less code
3.03 - 6/7/2006 - Fixed back button behaviour in IE -- thanks Terry!
3.03 - 6/7/2006 - Fixed infinite loop in Gecko browsers if unable to set designMode.
3.02a - 5/25/2006 - Changed default width to 500px
3.02a - 5/25/2006 - Changed defaults for cut, copy, paste, undo, redo, remove formatting commands to false
3.02a - 5/25/2006 - Fixed color palette positioning
3.02 - 5/25/2006 - Fixed image insert
3.02 - 5/25/2006 - Added ability to disable spellcheck button
3.02 - 5/25/2006 - updateRTEs() function no longer displayes error for pages calling this function with no RTEs
3.01a - 5/16/2006 - Fixed bug with color palette in IE
3.0 - 5/15/2006 - ColdFusion demo added
3.0 - 5/15/2006 - Major code cleanup.  RTEs are now instantiated with a class.  Toolbars and individual commands can now be enabled/disabled through the class.
3.0 - 5/15/2006 - Fixed bug with inserting links (link text was not automatically entered)
3.0 - 5/15/2006 - UTF-8 character set support added
3.0 - 5/15/2006 - Opera 9 support added
3.0 - 5/15/2006 - Partial support for Mac Safari 1.3+ (Bold, Italic, Underline, Left, Middle, & Right Align)
3.0 - 5/15/2006 - Implemented product versioning (starting at 3.0)
3/4/2005 - Updated insert link function to automatically place highlighted text in link text field
3/2/2005 - Replaced insert image icon
3/2/2005 - To satisfy a handful of people who have a problem with my "no license" policy I have added "// This code is public domain. Redistribution and use of this code, with or without modification, is permitted." to top of richtext.js.
3/1/2005 - Fixed display of buttons when submitting in view source mode
3/1/2005 - Fixed toggleHTMLSrc() function to update view source textbox with XHTML-converted code
3/1/2005 - Removed XHTML demo and included genXHTML option in initRTE function is standard demo

examples/demo.php
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--
// Cross-Browser Rich Text Editor
// Written by Kevin Roth (http://www.kevinroth.com/rte/)
// License: http://creativecommons.org/licenses/by/2.5/
//-->
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
	<title>Cross-Browser Rich Text Editor (RTE)</title>
	<meta http-equiv="content-type" content="text/html; charset=UTF-8" />
	<meta name="keywords" content="cross-browser rich text editor, rte, textarea, htmlarea, content management, cms, blog, internet explorer, firefox, safari, opera, netscape, konqueror" />
	<meta name="description" content="The cross-browser rich-text editor (RTE) is based on the designMode() functionality introduced in Internet Explorer 5, and implemented in Mozilla 1.3+ using the Mozilla Rich Text Editing API." />
	<meta name="author" content="Kevin Roth" />
	<meta name="ROBOTS" content="ALL" />
	<!-- html2xhtml.js written by Jacob Lee <letsgolee@lycos.co.kr> //-->
	<script language="JavaScript" type="text/javascript" src="../cbrte/html2xhtml.min.js"></script>
	<script language="JavaScript" type="text/javascript" src="../cbrte/richtext.min.js"></script>
</head>
<body>

<h2>Cross-Browser Rich Text Editor (RTE) Demo</h2>
<p>For more information, visit the <a href="http://www.kevinroth.com/rte/">Cross-Browser Rich Text Editor (RTE) home page</a>.</p>

<!-- START Demo Code -->
<form name="RTEDemo" action="demo.php" method="post" onsubmit="return submitForm();">
<script language="JavaScript" type="text/javascript">
<!--
function submitForm() {
	//make sure hidden and iframe values are in sync for all rtes before submitting form
	updateRTEs();
	
	return true;
}

//Usage: initRTE(imagesPath, includesPath, cssFile, genXHTML, encHTML)
initRTE("../cbrte/images/", "../cbrte/", "", true);
//-->
</script>
<noscript><p><b>Javascript must be enabled to use this form.</b></p></noscript>

<script language="JavaScript" type="text/javascript">
<!--
//build new richTextEditor
var rte1 = new richTextEditor('rte1');
<?php
//format content for preloading
if (!(isset($_POST["rte1"]))) {
	$content = "here's the " . chr(13) . "\"preloaded <b>content</b>\"";
	$content = rteSafe($content);
} else {
	//retrieve posted value
	$content = rteSafe($_POST["rte1"]);
}
?>
rte1.html = '<?=$content;?>';
//rte1.toggleSrc = false;
rte1.build();
//-->
</script>
<p><input type="submit" name="submit" value="Submit" /></p>
</form>
<?php
function rteSafe($strText) {
	//returns safe code for preloading in the RTE
	$tmpString = $strText;
	
	//convert all types of single quotes
	$tmpString = str_replace(chr(145), chr(39), $tmpString);
	$tmpString = str_replace(chr(146), chr(39), $tmpString);
	$tmpString = str_replace("'", "&#39;", $tmpString);
	
	//convert all types of double quotes
	$tmpString = str_replace(chr(147), chr(34), $tmpString);
	$tmpString = str_replace(chr(148), chr(34), $tmpString);
//	$tmpString = str_replace("\"", "\"", $tmpString);
	
	//replace carriage returns & line feeds
	$tmpString = str_replace(chr(10), " ", $tmpString);
	$tmpString = str_replace(chr(13), " ", $tmpString);
	
	return $tmpString;
}
?>
<!-- END Demo Code -->

</body>
</html>


cbrte/html2xhtml.js

/*
Html2Xhtml : Javascript Mini HTML to XHTML Parser 
--------------------------------------------------

Copyright 2006 All rights reserved. Jacob Lee <letsgolee@lycos.co.kr>

Free for noncommercial && commercial use.

Modified: 9/27/2010 Kevin Roth - updated to keep correct case on special characters such as Á, Ü, Ê etc
*/
function getXHTML(data) {
	return new Html2Xhtml(data).parse();
}
function Html2Xhtml(data) {
	this.data = data || "";
}
Html2Xhtml.prototype.setHTML = function (data) {
	this.data = data || this.data;
};
Html2Xhtml.prototype.parse = function () {
	var state = 0;
	var xhtml = "";
	var p = 0;
	var unget = false;
	var tagname = "";
	var attrname = "";
	var attrval = "";
	var quot = "";
	var data = this.data;
	var len = data.length;
	var phpval = "";
	var tagtype = 0;
	var insidepre = false;
	while (1) {
		if (p >= len && !unget) {
			return xhtml;
		}
		if (unget) {
			unget = false;
		} else {
			var c = data.substr(p++, 1);
		}
		switch (state) {
		  case 0:
			if (c == "<") {
				state = 1;
				break;
			}
			var cc = c.charCodeAt();
			if (Html2Xhtml.charEntities[cc]) {
				xhtml += "&" + Html2Xhtml.charEntities[cc] + ";";
			} else {
				xhtml += c;
			}
			break;
		  case 1:
			if (/[a-zA-Z]/.test(c)) {
				state = 2;
				tagtype = 1;
				tagname = c.toLowerCase();
				break;
			}
			if (c == "/") {
				state = 2;
				tagtype = -1;
				break;
			}
			if (c == "!") {
				if (data.substr(p, 2) == "--") {
					xhtml += "<!--";
					p += 2;
					state = 9;
					break;
				}
				xhtml += "<!";
				state = 10;
				break;
			}
			if (c == "?") {
				state = 11;
				xhtml += "<" + "?";
				break;
			}
			xhtml += "&lt;";
			unget = true;
			state = 0;
			break;
		  case 2:
			if (Html2Xhtml.isSpaceChar[c]) {
				xhtml += (!insidepre && tagtype > 0 && Html2Xhtml.hasNLBefore[tagname] && xhtml.length && xhtml.substr(xhtml.length - 1, 1) != "\n" ? "\n" : "") + (tagtype > 0 ? "<" : "</") + tagname;
				state = 3;
				break;
			}
			if (c == "/") {
				xhtml += (!insidepre && tagtype > 0 && Html2Xhtml.hasNLBefore[tagname] && xhtml.length && xhtml.substr(xhtml.length - 1, 1) != "\n" ? "\n" : "") + (tagtype > 0 ? "<" : "</") + tagname;
				if (data.substr(p, 1) != ">") {
					state = 3;
					break;
				}
				state = 4;
				break;
			}
			if (c == ">") {
				xhtml += (!insidepre && tagtype > 0 && Html2Xhtml.hasNLBefore[tagname] && xhtml.length && xhtml.substr(xhtml.length - 1, 1) != "\n" ? "\n" : "") + (tagtype > 0 ? "<" : "</") + tagname;
				unget = true;
				state = 4;
				break;
			}
			tagname += c.toLowerCase();
			break;
		  case 3:
			if (Html2Xhtml.isSpaceChar[c]) {
				break;
			}
			if (c == "/") {
				if (data.substr(p, 1) != ">") {
					break;
				}
				state = 4;
				break;
			}
			if (c == ">") {
				unget = true;
				state = 4;
				break;
			}
			attrname = c.toLowerCase();
			attrval = "";
			state = 5;
			break;
		  case 4:
			xhtml += (Html2Xhtml.isEmptyTag[tagname] ? " />" : ">") + (!insidepre && tagtype < 0 && Html2Xhtml.hasNLAfter[tagname] && p < len && data.substr(p, 1) != "\n" ? "\n" : "");
			if (tagtype > 0 && Html2Xhtml.dontAnalyzeContent[tagname]) {
				state = 13;
				attrname = attrval = quot = "";
				tagtype = 0;
				break;
			}
			if (tagname == "pre") {
				insidepre = !insidepre;
			}
			state = 0;
			tagname = attrname = attrval = quot = "";
			tagtype = 0;
			break;
		  case 5:
			if (Html2Xhtml.isSpaceChar[c]) {
				xhtml += " " + attrname;
				if (Html2Xhtml.isEmptyAttr[attrname]) {
					xhtml += "=\"" + attrname + "\"";
				}
				state = 3;
				break;
			}
			if (c == "/") {
				xhtml += " " + attrname;
				if (Html2Xhtml.isEmptyAttr[attrname]) {
					xhtml += "=\"" + attrname + "\"";
				}
				if (data.substr(p, 1) != ">") {
					state = 3;
					break;
				}
				state = 4;
				break;
			}
			if (c == ">") {
				xhtml += " " + attrname;
				if (Html2Xhtml.isEmptyAttr[attrname]) {
					xhtml += "=\"" + attrname + "\"";
				}
				unget = true;
				state = 4;
				break;
			}
			if (c == "=") {
				xhtml += " " + attrname + "=";
				state = 6;
				break;
			}
			if (c == "\"" || c == "'") {
				attrname += "?";
			} else {
				attrname += c.toLowerCase();
			}
			break;
		  case 6:
			if (Html2Xhtml.isSpaceChar[c]) {
				xhtml += (Html2Xhtml.isEmptyAttr[attrname] ? "\"" + attrname + "\"" : "\"\"");
				state = 3;
				break;
			}
			if (c == ">") {
				xhtml += (Html2Xhtml.isEmptyAttr[attrname] ? "\"" + attrname + "\"" : "\"\"");
				unget = true;
				state = 4;
				break;
			}
			if (c == "/" && data.substr(p, 1) == ">") {
				xhtml += (Html2Xhtml.isEmptyAttr[attrname] ? "\"" + attrname + "\"" : "\"\"");
				state = 4;
				break;
			}
			if (c == "\"" || c == "'") {
				quot = c;
				state = 8;
				break;
			}
			attrval = c;
			state = 7;
			break;
		  case 7:
			if (Html2Xhtml.isSpaceChar[c]) {
				xhtml += "\"" + Html2Xhtml.escapeQuot(attrval, "\"") + "\"";
				state = 3;
				break;
			}
			if (c == "/" && data.substr(p, 1) == ">") {
				xhtml += "\"" + Html2Xhtml.escapeQuot(attrval, "\"") + "\"";
				state = 4;
				break;
			}
			if (c == ">") {
				unget = true;
				xhtml += "\"" + Html2Xhtml.escapeQuot(attrval, "\"") + "\"";
				state = 4;
				break;
			}
			attrval += c;
			break;
		  case 8:
			if (c == quot) {
				xhtml += "\"" + Html2Xhtml.escapeQuot(attrval, "\"") + "\"";
				state = 3;
				break;
			}
			attrval += c;
			break;
		  case 9:
			if (c == "-" && data.substr(p, 2) == "->") {
				p += 2;
				xhtml += "-->";
				state = 0;
				break;
			}
			xhtml += c;
			break;
		  case 10:
			if (c == ">") {
				state = 0;
			}
			xhtml += c;
			break;
		  case 11:
			if (c == "'" || c == "\"") {
				quot = c;
				state = 12;
				break;
			}
			if (c == "?" && data.substr(p, 1) == ">") {
				state = 0;
				xhtml += "?" + ">";
				p++;
				break;
			}
			xhtml += c;
			break;
		  case 12:
			if (c == quot) {
				state = 11;
				xhtml += quot + Html2Xhtml.escapeQuot(phpval, quot) + quot;
				phpval = quot = "";
				break;
			}
			phpval += c;
			break;
		  case 13:
			if (c == "<" && data.substr(p, tagname.length + 1).toLowerCase() == "/" + tagname) {
				unget = true;
				state = 0;
				tagname = "";
				break;
			}
			if (tagname == "textarea") {
				xhtml += Html2Xhtml.escapeHTMLChar(c);
			} else {
				xhtml += c;
			}
			break;
		}
	}
	return xhtml;
};
Html2Xhtml.escapeQuot = function (str, quot) {
	if (!quot) {
		quot = "\"";
	}
	if (quot == "\"") {
		return str.replace(/"/ig, "\\\"");
	}
	return str.replace(/'/ig, "\\'");
};
Html2Xhtml.escapeHTMLChar = function (c) {
	if (c == "&") {
		return "&amp;";
	}
	if (c == "<") {
		return "&lt;";
	}
	if (c == ">") {
		return "&gt;";
	}
	var cc = c.charCodeAt();
	if (Html2Xhtml.charEntities[cc]) {
		return "&" + Html2Xhtml.charEntities[cc] + ";";
	} else {
		return c;
	}
};
Html2Xhtml.isSpaceChar = {" ":1, "\r":1, "\n":1, "\t":1};
Html2Xhtml.isEmptyTag = {"area":1, "base":1, "basefont":1, "br":1, "hr":1, "img":1, "input":1, "link":1, "meta":1, "param":1};
Html2Xhtml.isEmptyAttr = {"checked":1, "compact":1, "declare":1, "defer":1, "disabled":1, "ismap":1, "multiple":1, "noresize":1, "nosave":1, "noshade":1, "nowrap":1, "readonly":1, "selected":1};
Html2Xhtml.hasNLBefore = {"div":1, "p":1, "table":1, "tbody":1, "tr":1, "td":1, "th":1, "title":1, "head":1, "body":1, "script":1, "comment":1, "li":1, "meta":1, "h1":1, "h2":1, "h3":1, "h4":1, "h5":1, "h6":1, "hr":1, "ul":1, "ol":1, "option":1, "link":1};
Html2Xhtml.hasNLAfter = {"html":1, "head":1, "body":1, "p":1, "th":1, "style":1};
Html2Xhtml.dontAnalyzeContent = {"textarea":1, "script":1, "style":1};
Html2Xhtml.charEntities = {160:"nbsp", 161:"iexcl", 162:"cent", 163:"pound", 164:"curren", 165:"yen", 166:"brvbar", 167:"sect", 168:"uml", 169:"copy", 170:"ordf", 171:"laquo", 172:"not", 173:"shy", 174:"reg", 175:"macr", 176:"deg", 177:"plusmn", 178:"sup2", 179:"sup3", 180:"acute", 181:"micro", 182:"para", 183:"middot", 184:"cedil", 185:"sup1", 186:"ordm", 187:"raquo", 188:"frac14", 189:"frac12", 190:"frac34", 191:"iquest", 192:"Agrave", 193:"Aacute", 194:"Acirc", 195:"Atilde", 196:"Auml", 197:"Aring", 198:"AElig", 199:"Ccedil", 200:"Egrave", 201:"Eacute", 202:"Ecirc", 203:"Euml", 204:"Igrave", 205:"Iacute", 206:"Icirc", 207:"Iuml", 208:"ETH", 209:"Ntilde", 210:"Ograve", 211:"Oacute", 212:"Ocirc", 213:"Otilde", 214:"Ouml", 215:"times", 216:"Oslash", 217:"Ugrave", 218:"Uacute", 219:"Ucirc", 220:"Uuml", 221:"Yacute", 222:"THORN", 223:"szlig", 224:"agrave", 225:"aacute", 226:"acirc", 227:"atilde", 228:"auml", 229:"aring", 230:"aelig", 231:"ccedil", 232:"egrave", 233:"eacute", 234:"ecirc", 235:"euml", 236:"igrave", 237:"iacute", 238:"icirc", 239:"iuml", 240:"eth", 241:"ntilde", 242:"ograve", 243:"oacute", 244:"ocirc", 245:"otilde", 246:"ouml", 247:"divide", 248:"oslash", 249:"ugrave", 250:"uacute", 251:"ucirc", 252:"uuml", 253:"yacute", 254:"thorn", 255:"yuml", 338:"oelig", 339:"oelig", 352:"scaron", 353:"scaron", 376:"yuml", 710:"circ", 732:"tilde", 8194:"ensp", 8195:"emsp", 8201:"thinsp", 8204:"zwnj", 8205:"zwj", 8206:"lrm", 8207:"rlm", 8211:"ndash", 8212:"mdash", 8216:"lsquo", 8217:"rsquo", 8218:"sbquo", 8220:"ldquo", 8221:"rdquo", 8222:"bdquo", 8224:"dagger", 8225:"dagger", 8240:"permil", 8249:"lsaquo", 8250:"rsaquo", 8364:"euro", 402:"fnof", 913:"alpha", 914:"beta", 915:"gamma", 916:"delta", 917:"epsilon", 918:"zeta", 919:"eta", 920:"theta", 921:"iota", 922:"kappa", 923:"lambda", 924:"mu", 925:"nu", 926:"xi", 927:"omicron", 928:"pi", 929:"rho", 931:"sigma", 932:"tau", 933:"upsilon", 934:"phi", 935:"chi", 936:"psi", 937:"omega", 945:"alpha", 946:"beta", 947:"gamma", 948:"delta", 949:"epsilon", 950:"zeta", 951:"eta", 952:"theta", 953:"iota", 954:"kappa", 955:"lambda", 956:"mu", 957:"nu", 958:"xi", 959:"omicron", 960:"pi", 961:"rho", 962:"sigmaf", 963:"sigma", 964:"tau", 965:"upsilon", 966:"phi", 967:"chi", 968:"psi", 969:"omega", 977:"thetasym", 978:"upsih", 982:"piv", 8226:"bull", 8230:"hellip", 8242:"prime", 8243:"prime", 8254:"oline", 8260:"frasl", 8472:"weierp", 8465:"image", 8476:"real", 8482:"trade", 8501:"alefsym", 8592:"larr", 8593:"uarr", 8594:"rarr", 8595:"darr", 8596:"harr", 8629:"crarr", 8656:"larr", 8657:"uarr", 8658:"rarr", 8659:"darr", 8660:"harr", 8704:"forall", 8706:"part", 8707:"exist", 8709:"empty", 8711:"nabla", 8712:"isin", 8713:"notin", 8715:"ni", 8719:"prod", 8721:"sum", 8722:"minus", 8727:"lowast", 8730:"radic", 8733:"prop", 8734:"infin", 8736:"ang", 8743:"and", 8744:"or", 8745:"cap", 8746:"cup", 8747:"int", 8756:"there4", 8764:"sim", 8773:"cong", 8776:"asymp", 8800:"ne", 8801:"equiv", 8804:"le", 8805:"ge", 8834:"sub", 8835:"sup", 8836:"nsub", 8838:"sube", 8839:"supe", 8853:"oplus", 8855:"otimes", 8869:"perp", 8901:"sdot", 8968:"lceil", 8969:"rceil", 8970:"lfloor", 8971:"rfloor", 9001:"lang", 9002:"rang", 9426:"copy", 9674:"loz", 9824:"spades", 9827:"clubs", 9829:"hearts", 9830:"diams"};

